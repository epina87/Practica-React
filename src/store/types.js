export const AUTH_LOGIN = 'AUTH_LOGIN';
export const AUTH_LOGOUT = 'AUTH_LOGOUT';

export const ADVERTS_LOADED = 'ADVERTS_LOADED'
export const ADVERTS_CREATED = 'ADVERTS_CREATED'