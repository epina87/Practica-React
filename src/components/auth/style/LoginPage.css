html,body { 
	height: 100%; 
}

.global-container{
	height:100%;
	display: flex;
	align-items: center;
	justify-content: center;
	background-color: #f5f5f5;
}

form{
	padding-top: 10px;
	font-size: 14px;
	margin-top: 30px;
}

p{
    text-align: left;
    margin-left: 75px;
}



.card-title{ font-weight:300; }

.btn{
	font-size: 14px;
	margin-top:20px;
}


.login-form{ 
	width:330px;
	margin:20px;
}

.sign-up{
	text-align:center;
	padding:20px 0 0;
}

.alert{
	margin-bottom:-30px;
	font-size: 13px;
	margin-top:20px;
}

.btn,.btnDisabled{
	width: 130px;
	height: 40px;
	color: #fff;
	border-radius: 5px;

	font-family: 'Lato', sans-serif;
	font-weight: 500;
	background: transparent;
	cursor: pointer;
	transition: all 0.3s ease;
	position: relative;
	display: inline-block;
	 box-shadow:inset 2px 2px 2px 0px rgba(255,255,255,.5),
	 7px 7px 20px 0px rgba(0,0,0,.1),
	 4px 4px 5px 0px rgba(0,0,0,.1);
     outline: none;

	background: rgb(13, 152, 83);
background: linear-gradient(0deg, rgb(8, 146, 72) 0%, rgb(21, 99, 72) 100%);

  line-height: 42px;
  padding: 0;
  border: none;

}

.btnDisabled{
    background: rgb(85, 112, 122);
    cursor: default;
}

.btn:before,
.btn:after {
  position: absolute;
  content: "";
  right: 0;
  top: 0;
   background: rgb(117, 219, 178);
  transition: all 0.3s ease;
}
.btn:before {
  height: 0%;
  width: 2px;
}
.btn:after {
  width: 0%;
  height: 2px;
}
.btn:hover{
   background: transparent;
  	box-shadow: none;
	  color: black;
}
.btn:hover:before {
  height: 100%;
}
.btn:hover:after {
  width: 100%;
}

.error{
    color: red;
    font-size: 20px;
}



