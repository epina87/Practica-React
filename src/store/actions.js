//import { getAdvertsList, getTags } from '../components/adverts/service';

import {
  areAdvertsLoaded,
  areTagsLoaded,
  getAdvert,
  getAdvertsNotDelete,
} from './selectors';
import {

  ADVERTS_LOADED_FAILURE,
  ADVERTS_LOADED_REQUEST,
  ADVERTS_LOADED_SUCCESS,
  ADVERT_CREATED_FAILURE,
  ADVERT_CREATED_REQUEST,
  ADVERT_CREATED_SUCCESS,
  ADVERT_DELETE_FAILURE,
  ADVERT_DELETE_REQUEST,
  ADVERT_DELETE_SUCCESS,
  ADVERT_LOADED_FAILURE,
  ADVERT_LOADED_REQUEST,
  ADVERT_LOADED_SUCCESS,
  AUTH_LOGIN_FAILURE,
  AUTH_LOGIN_REQUEST,
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGOUT,

  TAGS_LOADED_FAILURE,
  TAGS_LOADED_REQUEST,
  TAGS_LOADED_SUCCESS,
  UI_RESET_ERROR,
} from './types';

export const authLoginRequest = () => ({
  type: AUTH_LOGIN_REQUEST,
});

export const authLoginSuccess = () => ({
  type: AUTH_LOGIN_SUCCESS,
});

export const authLoginFailure = error => ({
  type: AUTH_LOGIN_FAILURE,
  error: true,
  payload: error,
});

export const authLogin =
  (credentials, saveSession) =>
  async (dispatch, _getState, { service, router }) => {
    dispatch(authLoginRequest());
    try {
      await service.auth.login(credentials, saveSession);
      dispatch(authLoginSuccess());
      const to = router.state?.location?.state?.from?.pathname || '/';
      
        router.navigate(to);
    } catch (error) {
      dispatch(authLoginFailure(error));
      
    }
  };

export const authLogout = () => ({
  type: AUTH_LOGOUT,
});

export const advertsLoadedRequest = () => ({
  type: ADVERTS_LOADED_REQUEST,
});

export const advertsLoadedSuccess = adverts => ({
  type: ADVERTS_LOADED_SUCCESS,
  payload: adverts,
});

export const advertsLoadedFailure = error => ({
  type: ADVERTS_LOADED_FAILURE,
  error: true,
  payload: error,
});

export const advertsLoaded =
  () =>
  async (dispatch, getState, { service}) => {

    
    if (areAdvertsLoaded(getState())) {
      return;
    }

    dispatch(advertsLoadedRequest());
    console.log(service)
    try {
      const adverts = await service.adverts.getAdvertsList();
  
      dispatch(advertsLoadedSuccess(adverts));
    } catch (error) {
      dispatch(advertsLoadedFailure(error));
      throw error;
    }
  };

export const advertLoad =
  advertId =>
  async (dispatch, getState, { service }) => {
    const isLoaded = getAdvert(advertId)(getState());

    if (isLoaded) {
      return;
    }

    dispatch(advertLoadedRequest());
    try {
      const advert = await service.adverts.getAdvert(advertId);
      dispatch(advertLoadedSuccess(advert));
    } catch (error) {
      dispatch(advertLoadedFailure(error));  
    }
  };

export const advertLoadedRequest = () => ({
  type: ADVERT_LOADED_REQUEST,
});

export const advertLoadedSuccess = advert => ({
  type: ADVERT_LOADED_SUCCESS,
  payload: advert,
});

export const advertLoadedFailure = error => ({
  type: ADVERT_LOADED_FAILURE,
  error: true,
  payload: error,
});

export const tagsLoadedSuccess = tags => ({
  type: TAGS_LOADED_SUCCESS,
  payload: tags,
});

export const tagsLoadedRequest = () => ({
  type: TAGS_LOADED_REQUEST,
});

export const tagsLoadedFailure = error => ({
  type: TAGS_LOADED_FAILURE,
  error: true,
  payload: error,
});

export const tagsLoaded =
  () =>
  async (dispatch, getState, { service }) => {
    if (areTagsLoaded(getState())) {
      return;
    }
    dispatch(tagsLoadedRequest());
    
    try {
      const tags = await service.adverts.getTags();
      dispatch(tagsLoadedSuccess(tags));
    } catch (error) {
      dispatch(tagsLoadedFailure(error));
      throw error;
    }
  };

export const uiResetError = () => ({
  type: UI_RESET_ERROR,
});

export const advertCreated =
  advert =>
  async (dispatch, _getState, { service, router}) => {
    dispatch(advertCreatedRequest());
    try {
      const createdAdvert = await service.adverts.createAdvert(advert, {
        headers: { 'content-type': 'multipart/form-data' },
      });

      dispatch(advertCreatedSuccess(createdAdvert));
      
      router.navigate(`/adverts/${createdAdvert.id}`);
      
    } catch (error) {
      dispatch(advertCreatedFailure(error));
      
    }
  };

export const advertCreatedRequest = () => ({
  type: ADVERT_CREATED_REQUEST,
});

export const advertCreatedSuccess = advert => ({
  type: ADVERT_CREATED_SUCCESS,
  payload: advert,
});

export const advertCreatedFailure = error => ({
  type: ADVERT_CREATED_FAILURE,
  error: true,
  payload: error,
});

export const advertDelete =
  advertId =>
  async (dispatch, getState, { service, router }) => {
    dispatch(advertDeleteRequest());
    try {
      await service.adverts.deleteAdvert(advertId);
      const adverts = getAdvertsNotDelete(advertId)(getState());

      dispatch(advertDeleteSuccess(adverts));
      router.navigate('/adverts');
    } catch (error) {
      dispatch(advertDeleteFailure(error));
      router.navigate('/404');
      
    }
  };

export const advertDeleteRequest = () => ({
  type: ADVERT_DELETE_REQUEST,
});

export const advertDeleteSuccess = advertId => ({
  type: ADVERT_DELETE_SUCCESS,
  payload: advertId,
});

export const advertDeleteFailure = error => ({
  type: ADVERT_DELETE_FAILURE,
  error: true,
  payload: error,
});
