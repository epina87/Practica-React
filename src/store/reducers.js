import {

  ADVERTS_LOADED_SUCCESS,
  ADVERT_CREATED_SUCCESS,
  ADVERT_DELETE_SUCCESS,
  ADVERT_LOADED_SUCCESS,

  AUTH_LOGIN_SUCCESS,
  AUTH_LOGOUT,

  TAGS_LOADED_SUCCESS,
  UI_RESET_ERROR,
} from './types';

const defaultState = {
  auth: false,
  adverts: {
    areLoaded: false,
    data: [],
  },
  tags: { areLoaded: false, data: [] },
  ui: {
    isLoading: false,
    error: null,
  },
};

export function auth(state = defaultState.auth, action) {
  switch (action.type) {
    case AUTH_LOGIN_SUCCESS:
      return true;
    case AUTH_LOGOUT:
      return false;
    default:
      return state;
  }
}

export function adverts(state = defaultState.adverts, action) {
  switch (action.type) {
    case ADVERTS_LOADED_SUCCESS:
      return { areLoaded: true, data: action.payload };
    case ADVERT_LOADED_SUCCESS:
      return { ...state, data: [action.payload] };
    case ADVERT_CREATED_SUCCESS:
        return {...state, data:[...state.data,action.payload]}
    case ADVERT_DELETE_SUCCESS:
        return {...state, data: action.payload}

    default:
      return state;
  }
}

export function tags(state = defaultState.tags, action) {
  switch (action.type) {
    case TAGS_LOADED_SUCCESS:
      return { areLoaded: true, data: action.payload };
    default:
      return state;
  }
}

export function ui(state = defaultState.ui, action) {
  if (action.error) {
    return { isLoading: false, error: action.payload };
  }

  if (/_REQUEST$/.test(action.type)) {
    return { isLoading: true, error: null };
  }

  if (/_SUCCESS$/.test(action.type)) {
    return { isLoading: false, error: null };
  }

  if (action.type === UI_RESET_ERROR) {
    return { ...state, error: null };
  }

  return state;
}
