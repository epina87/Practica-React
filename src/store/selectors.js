import { getAdvertsList } from "../components/adverts/service"

export const getIsLogged  = state => state.auth

 //export const getIsAdverts = state => state.adverts.areLoaded ? state.adverts.data : [] 

 export const getIsAdverts = state => state.adverts.data


export const getIsTags    = state => state.tags.data

export const getAdvert = advertId => state => state.adverts.data.find(advert => advert.id === advertId) 

// export const getAdvert =  advertId => state => {

//     console.log(state)
//     // const advert = getIsAdverts(state)
//     // console.log('selectors 1',advert)
//     // advert.find(advert => advert.id === advertId) 

//     // console.log('selectors 2',advert)
    
    
// }

export const getAdvertsNotDelete = advertId => state => state.adverts.data.filter(advert => advert.id !== advertId)


export const getUi = state => state.ui


export const areAdvertsLoaded = state => state.adverts.areLoaded;
export const areTagsLoaded = state => state.tags.areLoaded;