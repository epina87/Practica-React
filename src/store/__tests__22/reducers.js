import { authLoginSuccess, authLogoutSucces } from '../actions';
import { auth, defaultState } from '../reducers';

describe('auth', () => {
  test('should manage "AUTH_LOGIN_SUCCESS" action', () => {
    const state = defaultState.auth;
    const action = authLoginSuccess();

    const newState = auth(state, action);
    expect(newState).toBe(true);
  });

  test('should manage "AUTH_LOGOUT" action', () => {
    const state = defaultState.auth;
    const action = authLogoutSucces();

    const newState = auth(state, action);
    expect(newState).toBe(false);
  });

  test('should manage any other action when state is not undefined', () => {
    const state = true;
    const action = { type: 'ANY' };

    const newState = auth(state, action);
    expect(newState).toBe(state);
  });
});
