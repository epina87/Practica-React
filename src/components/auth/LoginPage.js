import { useState } from 'react';
import './style/LoginPage.css';
import { useDispatch, useSelector } from 'react-redux';
import {
  authLogin,
  uiResetError,
} from '../../store/actions';
import { getUi } from '../../store/selectors';

function LoginPage() {
  const dispatch = useDispatch();

  const { isLoading, error } = useSelector(getUi);

  const [credential, setCredentials] = useState({
    email: '',
    password: '',
  });

  const resetError = () => {
    dispatch(uiResetError());
  };

  const [saveSession, setSaveSession] = useState(false);

  const handlechecked = event => {
    setSaveSession(event.target.checked);
  };

  const handleSubmit =  event => {
    event.preventDefault();
    dispatch(authLogin(credential, saveSession))
  };

  const handleChange = event => {
    setCredentials({ ...credential, [event.target.name]: event.target.value });
  };

  const buttonDisabled = isLoading || !credential.email || !credential.password;
  const btnClass = !buttonDisabled ? 'btn' : 'btnDisabled';

  return (
    <div className="global-container">
      <div className="card login-form">
        <div className="card-body">
          <h1 className="card-title text-center">New User Register</h1>
          <div className="card-text">
            <form id="createUser" onSubmit={handleSubmit}>
              <p>
                <label>Email
                <br />
                <input
                  type="email"
                  name="email"
                  id="email"
                  required
                  onChange={handleChange}
                  value={credential.email}
                />
                </label>
              </p>
              <p>
                <label>Password
                <br />
                <input
                  type="password"
                  name="password"
                  id="password"
                  required
                  onChange={handleChange}
                  value={credential.password}
                />
                </label>
              </p>
              <button
                type="submit"
                className={btnClass}
                disabled={buttonDisabled}
              >
                Sign in
              </button>

              <p>
                <input type="checkbox" onChange={handlechecked} /> you want to
                save the session
              </p>
              {error && (
                <div className="error" onClick={resetError}>
                  {' '}
                  {error.message}{' '}
                </div>
              )}
            </form>
          </div>
        </div>
      </div>
    </div>
  );
}

export default LoginPage;
