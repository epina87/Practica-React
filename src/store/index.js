import { createStore, combineReducers, applyMiddleware } from 'redux';
import { composeWithDevTools } from '@redux-devtools/extension';
import thunk from 'redux-thunk';
import * as auth from '../components/auth/service';
import * as adverts from '../components/adverts/service';

import * as reducers from './reducers';
import * as actionCreators from './actions';

const reducer = combineReducers(reducers);

const composeEnhancers = composeWithDevTools({
  actionCreators,
});

// const thunk = store => next => action =>{
//     if(typeof action === 'function'){
//         return action(store.dispatch, store.getState)
//     }
//     return next(action)
// }

const failureRedirects = (router,redirectsMap) => store => next => action => {
  const result = next(action);

  if (action.error) {
    const redirect = redirectsMap[action.payload.status]
    if (redirect){
        router.navigate(redirect);
    }
  }
  return result;
};

export default function configureStore(preloadedState, { router }) {
  const middleware = [
    thunk.withExtraArgument({ service: { auth, adverts }, router }),
    failureRedirects(router,{401:'/login',404:'/404'}),
  ];
  const store = createStore(
    reducer,
    preloadedState,
    composeEnhancers(applyMiddleware(...middleware)),
  );
  return store;
}
