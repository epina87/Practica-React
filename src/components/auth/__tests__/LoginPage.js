import { render, screen } from '@testing-library/react';
import LoginPage from '../LoginPage';
import { Provider } from 'react-redux';
import { defaultState } from '../../../store/reducers';
import { authLogin, uiResetError } from '../../../store/actions';
import userEvent from '@testing-library/user-event';

jest.mock('../../../store/actions');


describe('LoginPage', () => {
  const renderComponent = (error = null) => {
    const store = {
      getState: () => {
        const state = defaultState;
        state.ui.error = error
        return state;
      },
      subscribe: () => {},
      dispatch: () => {},
    };
    return render(
      <Provider store={store}>
        <LoginPage />
      </Provider>,
    );
  };



  test('snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test('should dispatch authLogin action', () => {
    const email = 'email';
    const password = '12345';
    const saveSession = false;

    renderComponent();

    const emailInput = screen.getByLabelText(/Email/);
    const passwordInput = screen.getByLabelText(/Password/);

    const submitButton = screen.getByRole('button');
    

    expect(submitButton).toBeDisabled();

    userEvent.type(emailInput,email)
    userEvent.type(passwordInput,password)


    expect(submitButton).toBeEnabled();
    userEvent.click(submitButton);

    expect(authLogin).toHaveBeenCalledWith({ email, password }, saveSession);

  });

   test('should display an error', () => {
    const error = {message: 'Unauthorized'}
    renderComponent(error);

    const errorElement = screen.getByText(error.message)
    expect(errorElement).toBeInTheDocument()


    userEvent.click(errorElement)

    expect(uiResetError).toHaveBeenCalled()
  });

  test('should checked save session',()=>{

    renderComponent();

    const chekedButton = screen.getByRole('checkbox');

     expect(chekedButton).not.toBeChecked();
     userEvent.click(chekedButton)
     expect(chekedButton).toBeChecked();
  });
});
