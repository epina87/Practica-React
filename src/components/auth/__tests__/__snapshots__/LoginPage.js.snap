// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage snapshot 1`] = `
<div>
  <div
    class="global-container"
  >
    <div
      class="card login-form"
    >
      <div
        class="card-body"
      >
        <h1
          class="card-title text-center"
        >
          New User Register
        </h1>
        <div
          class="card-text"
        >
          <form
            id="createUser"
          >
            <p>
              <label>
                Email
                <br />
                <input
                  id="email"
                  name="email"
                  required=""
                  type="email"
                  value=""
                />
              </label>
            </p>
            <p>
              <label>
                Password
                <br />
                <input
                  id="password"
                  name="password"
                  required=""
                  type="password"
                  value=""
                />
              </label>
            </p>
            <button
              class="btnDisabled"
              disabled=""
              type="submit"
            >
              Sign in
            </button>
            <p>
              <input
                type="checkbox"
              />
               you want to save the session
            </p>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
